apps/rebar/src/rebar_completion.erl

```diff
@@ -0,0 +1,34 @@
+-module(rebar_completion).
+
+-export([generate/2]).
+
+-type arg_type() :: atom | binary | boolean | float | interger | string.
+
+-type cmpl_arg() :: #{short => char() | undefined,
+                      long => string() | undefined,
+                      type => arg_type(),
+                      help => string()}.
+
+-type cmpl_cmd() :: #{name := string(),
+                      help := string() | undefined,
+                      args := [cmpl_arg()],
+                      cmds => [cmpl_cmd()]}.
+
+-type cmpl_opts() :: #{aliases => [string()],
+                       file => file:filename(),
+                       %% TODO support fish and maybe some more shells
+                       shell => bash | zsh}.
+-export([prelude/1]).
+
+-export_type([cmpl_opts/0, cmpl_cmd/0, cmpl_arg/0]).
+
+-callback generate([cmpl_cmd()], cmpl_opts()) -> iolist().
+
+-spec generate([cmpl_cmd()], cmpl_opts()) -> string().
+generate(Commands, #{shell:=bash}=CmplOpts) ->
+    rebar_completion_bash:generate(Commands,CmplOpts);
+generate(Commands, #{shell:=zsh}=CmplOpts) ->
+    rebar_completion_zsh:generate(Commands,CmplOpts).
```

Comment:
Looks like this call is failing in Dialyzer here:

```
apps/rebar/src/rebar_completion.erl
Line 27 Column 2: Invalid type specification for function rebar_completion:generate/2. The success typing is ([#{'args':=[map()], 'help':='undefined' | string(), 'name':=string(), 'cmds'=>[map()]}],#{'shell':='bash' | 'zsh', 'aliases'=>[string()], 'file'=>string()}) -> [[[any()] | char()],...] but the spec is ([cmpl_cmd()],cmpl_opts()) -> string()
```
Of significance here I think is that the return value is supposed to be a string, but you're generating a deeply nested list of possibly non-string elements. I'm not sure where it finds the `any()` non-string element in there (maybe it just gives up at a certain depth), but you're likely going to want to either change the return type to unicode data (as exported by the `unicode` module) or `chardata()` or any other recursive type, or flatten the value with the `unicode` module into a regular string.

The `iolist()` spec specified in the completion-specific modules are at the very least clashing here.